drivers/sensor/ite_vcmp_it8xxx2/vcmp_ite_it8xxx2.c

```diff
@@ -0,0 +1,398 @@
+	volatile uint8_t *reg_vcmpsts = (uint8_t *)config->reg_vcmpsts;
+	volatile uint8_t *reg_vcmpsts2 = (uint8_t *)config->reg_vcmpsts2;
+	int idx, status;
+
+	/* Find out which voltage comparator triggered */
+	status = *reg_vcmpsts & VCMP_REG_MASK;
+	status |= (*reg_vcmpsts2 & VCMP_REG_MASK) << 3;
+
+	for (idx = VCMP_CHANNEL_0; idx < VCMP_CHANNEL_CNT; idx++) {
+		if (status & BIT(idx)) {
+			/* Trigger specific channel callback function in work queue */
+			if (vcmp_work_addr[idx]) {
+				k_work_submit((struct k_work *) vcmp_work_addr[idx]);
+			}
+			/* W/C voltage comparator specific channel interrupt status */
+			clear_vcmp_status(dev, idx);
+		}
+	}
+
+	/* W/C voltage comparator irq interrupt status */
+	ite_intc_isr_clear(config->irq);
+}
+
+static int vcmp_it8xxx2_init(const struct device *dev)
+{
+	const struct vcmp_it8xxx2_config *const config = dev->config;
+	struct vcmp_it8xxx2_data *const data = dev->data;
+	volatile uint8_t *reg_vcmpxctl = (uint8_t *)config->reg_vcmpxctl;
+	volatile uint8_t *reg_vcmpxcselm = (uint8_t *)config->reg_vcmpxcselm;
+	volatile uint8_t *reg_vcmpscp = (uint8_t *)config->reg_vcmpscp;
+
+	/* Disable voltage comparator specific channel before init */
+	vcmp_enable(dev, 0);
+
+	/*
+	 * ADC channel signal output to voltage comparator,
+	 * so we need to set ADC channel to alternate mode first.
+	 */
+	adc_channel_setup(config->adc, &data->adc_ch_cfg);
+
+	/* Select which ADC channel output voltage into comparator */
+	if (data->adc_ch_cfg.channel_id <= 7) {
+		/* ADC channel 0~7 map to value 0x0~0x7 */
+		*reg_vcmpxctl |= data->adc_ch_cfg.channel_id & VCMP_REG_MASK;
+		*reg_vcmpxcselm &= ~IT8XXX2_VCMP_VCMPXCSELM;
+	} else {
+		/* ADC channel 13~16 map to value 0x8~0xb */
+		*reg_vcmpxctl |= (data->adc_ch_cfg.channel_id - 5) & VCMP_REG_MASK;
+		*reg_vcmpxcselm |= IT8XXX2_VCMP_VCMPXCSELM;
+	}
+
+	/* Set minimum scan period for "all voltage comparator channel" */
+	if (*reg_vcmpscp > config->scan_period) {
+		*reg_vcmpscp = config->scan_period;
+	}
+
+	/* Data must keep device reference for worker handler */
+	data->vcmp = dev;
+
+	/* Init and set worker queue to enable notifications */
+	k_work_init(&data->work, it8xxx2_vcmp_trigger_work_handler);
+	vcmp_work_addr[config->vcmp_ch] = (uint32_t) &data->work;
+
+	/* Set threshold and comparison if set on device tree */
+	if ((config->threshold_mv != IT8XXX2_VCMP_UNDEFINED) &&
+	     (config->comparison != IT8XXX2_VCMP_UNDEFINED)) {
+		enum sensor_attribute attr;
+		struct sensor_value val;
+
+		if (config->comparison == IT8XXX2_VCMP_LESS_OR_EQUAL) {
+			attr = SENSOR_ATTR_LOWER_VOLTAGE_THRESH;
+		} else {
+			attr = SENSOR_ATTR_UPPER_VOLTAGE_THRESH;
+		}
+
+		val.val1 = config->threshold_mv;
+		val.val2 = 0;
+
+		vcmp_ite_it8xxx2_attr_set(dev, SENSOR_CHAN_VOLTAGE, attr, &val);
+	}
+
+	ite_intc_isr_clear(config->irq);
+
+	irq_connect_dynamic(config->irq, 0,
+			    (void (*)(const void *))vcmp_it8xxx2_isr,
+			    (const void *)dev, 0);
```

Comment:
why is dynamic connection needed?

Reply:
Because there are six vcmp channels, compiler doesn't know the arguments of which vcmp channel input at build time. If change to IRQ_CONNECT, it shows the error: initializer element is not constant.

Reply:
isn't it possible to do something like this https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_gd32.c#L694 ?

Reply:
Done